File: kirun-js/__tests__/engine/function/system/array/CopyTest.ts
import { AbstractArrayFunction } from '../../../../../src/engine/function/system/array/AbstractArrayFunction';
import { Copy } from '../../../../../src/engine/function/system/array/Copy';
import { FunctionOutput } from '../../../../../src/engine/model/FunctionOutput';
import { FunctionExecutionParameters } from '../../../../../src/engine/runtime/FunctionExecutionParameters';

import { MapUtil } from '../../../../../src/engine/util/MapUtil';

import { KIRunFunctionRepository, KIRunSchemaRepository } from '../../../../../src';

test('Compy Test', async () => {
    let copy: Copy = new Copy();

    let source: any[] = [1, 2, 3, 4, 5];

    let fep1: FunctionExecutionParameters = new FunctionExecutionParameters(
        new KIRunFunctionRepository(),
        new KIRunSchemaRepository(),
    ).setArguments(
        MapUtil.of(
            AbstractArrayFunction.PARAMETER_ARRAY_SOURCE.getParameterName(),
            source as any,
            AbstractArrayFunction.PARAMETER_INT_SOURCE_FROM.getParameterName(),
            2,
            AbstractArrayFunction.PARAMETER_INT_LENGTH.getParameterName(),
            4,
        ),
    );

    await expect(copy.execute(fep1)).rejects.toThrow();

    source.push(6);

    let result: any[] = [3, 4, 5, 6];

    let fep: FunctionExecutionParameters = new FunctionExecutionParameters(
        new KIRunFunctionRepository(),
        new KIRunSchemaRepository(),
    ).setArguments(
        MapUtil.of(
            Copy.PARAMETER_ARRAY_SOURCE.getParameterName(),
            source as any,
            Copy.PARAMETER_INT_SOURCE_FROM.getParameterName(),
            2,
            Copy.PARAMETER_INT_LENGTH.getParameterName(),
            4,
        ),
    );

    let fo: FunctionOutput = await copy.execute(fep);
    expect(fo.allResults()[0].getResult().get(Copy.EVENT_RESULT_NAME)).toStrictEqual(result);

    source = new Array();
    source.push({ name: 'Kiran' });
    source.push({ name: 'Kumar' });

    fep = new FunctionExecutionParameters(
        new KIRunFunctionRepository(),
        new KIRunSchemaRepository(),
    ).setArguments(
        MapUtil.of(
            Copy.PARAMETER_ARRAY_SOURCE.getParameterName(),
            source as any,
            Copy.PARAMETER_INT_SOURCE_FROM.getParameterName(),
            2,
            Copy.PARAMETER_INT_LENGTH.getParameterName(),
            4,
        ),
    );

    result = new Array();
    result.push({ name: 'Kiran' });
    result.push({ name: 'Kumar' });

    fep = new FunctionExecutionParameters(
        new KIRunFunctionRepository(),
        new KIRunSchemaRepository(),
    ).setArguments(MapUtil.of(Copy.PARAMETER_ARRAY_SOURCE.getParameterName(), source));

    fo = await copy.execute(fep);

    expect(fo.allResults()[0].getResult().get(Copy.EVENT_RESULT_NAME)).toStrictEqual(result);
    expect(source[0] == result[0]).toBeFalsy();

    fep = new FunctionExecutionParameters(
        new KIRunFunctionRepository(),
        new KIRunSchemaRepository(),
    ).setArguments(
        MapUtil.of(
            Copy.PARAMETER_ARRAY_SOURCE.getParameterName(),
            source as any,
            Copy.PARAMETER_BOOLEAN_DEEP_COPY.getParameterName(),
            false,
        ),
    );

    fo = await copy.execute(fep);

    result = fo.allResults()[0].getResult().get(Copy.EVENT_RESULT_NAME);

    expect(source[0] == result[0]).toBeTruthy();
});

File: kirun-java/src/test/java/com/fincity/nocode/kirun/engine/function/system/array/CopyTest.java
package com.fincity.nocode.kirun.engine.function.system.array;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertNotSame;

import java.util.Map;

import org.junit.jupiter.api.Test;

import com.fincity.nocode.kirun.engine.exception.KIRuntimeException;
import com.fincity.nocode.kirun.engine.repository.reactive.KIRunReactiveFunctionRepository;
import com.fincity.nocode.kirun.engine.repository.reactive.KIRunReactiveSchemaRepository;
import com.fincity.nocode.kirun.engine.runtime.reactive.ReactiveFunctionExecutionParameters;
import com.google.gson.JsonArray;
import com.google.gson.JsonObject;
import com.google.gson.JsonPrimitive;

import reactor.test.StepVerifier;

class CopyTest {

	@Test
	void test() {
		Copy copy = new Copy();

		JsonArray source = new JsonArray();

		source.add(1);
		source.add(2);
		source.add(3);
		source.add(4);
		source.add(5);

		final ReactiveFunctionExecutionParameters fep1 = new ReactiveFunctionExecutionParameters(
				new KIRunReactiveFunctionRepository(), new KIRunReactiveSchemaRepository())
				.setArguments(Map.of(Copy.PARAMETER_ARRAY_SOURCE.getParameterName(), source,
						Copy.PARAMETER_INT_SOURCE_FROM.getParameterName(), new JsonPrimitive(2),
						Copy.PARAMETER_INT_LENGTH.getParameterName(), new JsonPrimitive(4)));

		StepVerifier.create(copy.execute(fep1))
				.expectError(KIRuntimeException.class)
				.verify();

		source.add(6);

		JsonArray result = new JsonArray();
		result.add(3);
		result.add(4);
		result.add(5);
		result.add(6);

		ReactiveFunctionExecutionParameters fep = new ReactiveFunctionExecutionParameters(
				new KIRunReactiveFunctionRepository(),
				new KIRunReactiveSchemaRepository())
				.setArguments(Map.of(Copy.PARAMETER_ARRAY_SOURCE.getParameterName(), source,
						Copy.PARAMETER_INT_SOURCE_FROM.getParameterName(), new JsonPrimitive(2),
						Copy.PARAMETER_INT_LENGTH.getParameterName(), new JsonPrimitive(4)));

		final JsonArray finResult = result;
		StepVerifier.create(copy.execute(fep))
				.expectNextMatches(fo -> {
					return finResult.equals(fo.allResults()
							.get(0)
							.getResult()
							.get(Copy.EVENT_RESULT_NAME));
				}).verifyComplete();

		source = new JsonArray();

		JsonObject obj = new JsonObject();
		obj.addProperty("name", "Kiran");
		source.add(obj);

		obj = new JsonObject();
		obj.addProperty("name", "Kumar");
		source.add(obj);

		fep = new ReactiveFunctionExecutionParameters(new KIRunReactiveFunctionRepository(),
				new KIRunReactiveSchemaRepository())
				.setArguments(Map.of(Copy.PARAMETER_ARRAY_SOURCE.getParameterName(),
						source, Copy.PARAMETER_INT_SOURCE_FROM.getParameterName(), new JsonPrimitive(2),
						Copy.PARAMETER_INT_LENGTH.getParameterName(), new JsonPrimitive(4)));

		result = new JsonArray();

		obj = new JsonObject();
		obj.addProperty("name", "Kiran");
		result.add(obj);

		obj = new JsonObject();
		obj.addProperty("name", "Kumar");
		result.add(obj);

		fep = new ReactiveFunctionExecutionParameters(new KIRunReactiveFunctionRepository(),
				new KIRunReactiveSchemaRepository())
				.setArguments(Map.of(Copy.PARAMETER_ARRAY_SOURCE.getParameterName(), source));

		final JsonArray finResult1 = result;
		StepVerifier.create(copy.execute(fep))
				.expectNextMatches(fo -> {
					return finResult1.equals(fo.allResults()
							.get(0)
							.getResult()
							.get(Copy.EVENT_RESULT_NAME));
				}).verifyComplete();

		assertNotSame(source.get(0), result.get(0));

		fep = new ReactiveFunctionExecutionParameters(new KIRunReactiveFunctionRepository(),
				new KIRunReactiveSchemaRepository())
				.setArguments(Map.of(Copy.PARAMETER_ARRAY_SOURCE.getParameterName(),
						source, Copy.PARAMETER_BOOLEAN_DEEP_COPY.getParameterName(), new JsonPrimitive(false)));

		StepVerifier.create(copy.execute(fep))
				.expectNextMatches(fo -> {
					return finResult1.equals(fo.allResults()
							.get(0)
							.getResult()
							.get(Copy.EVENT_RESULT_NAME));
				}).verifyComplete();

		assertEquals(source.get(0), result.get(0));
	}

}
